### ISSUE TITLE
HPI_LOGS envvar no longer works

### ISSUE THREAD

**Issue by seanbreckenridge** [karlicoss/HPI#305]
Seems [got removed when hpi switched to colorlog](https://github.com/karlicoss/HPI/commit/661714f1d9d56f8c422fb2a12578ee04ccc96329#diff-390a6ba268d0a76aad1f6a97ff3626b88731f24559ad6ea668c67ce0e956886cL41-L44), is there plans to move it to some part of the `my.config.core`?

Otherwise I like having the ability to quickly introspect the cachew hash using it, can make a PR to re-add

**Comment by seanbreckenridge** [comment]
Ah, I guess theres the `LOGGING_LEVEL_` prefixes now, but not sure if it lets you configure everything. The `--debug` flag in the cli hasnt changed [here](https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/__main__.py#L599-L601)

Not sure if  `LOGGING_LEVEL`  should be added as a environment variable, or maybe `LOGGING_LEVEL_GLOBAL`?

**Comment by karlicoss** [comment]
Ah sorry I've been meaning to document but got distracted! I'll take a look later today!

**Comment by seanbreckenridge** [comment]
Am glad to PR some changes the google takeout parser/browserexport modules which previously used the `HPI_LOGS` to update those packages' log levels as well (I'd probably just use LazyLogger to create the logger for each module and then check the loggers computed level to)

Though, not sure what to do about the `--debug` [flag here](https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/__main__.py#L591), I guess could just do a:

```
import .logging as log_module
log_module.DEFAULT_LEVEL = 'DEBUG'
```

Alternatively, could change that line in the logging module to `DEFAULT_LEVEL = os.environ.get("LOGGING_LEVEL", "INFO")`, that would act pretty similarly to `HPI_LOGS` in the past (could even add a check for `HPI_LOGS` to keep backwards compatibility)

**Comment by karlicoss** [comment]
So there are a few things here (none of which I properly documented, sorry :see_no_evil: )

**HPI changes**
- added a helper to get logging level from an env variable, so it's easy to adjust per-module
  https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/logging.py#L78-L81
- only set logger level if it's not been set before
  https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/logging.py#L104-L106
  this is more useful if you call HPI stuff in a script or other tool, that way you can configure loggers first and then call HPI stuff
  e.g. `logging.getLogger('ghexport.dal').setLevel(logging.ERROR)`
- there is an experimental `ENLIGHTEN_ENABLE` flag, which does some magic to render a progress bar instead of dumping tons of logger messages
  ![Peek 2023-09-06 23-29 hpi enlighten demo](https://github.com/karlicoss/HPI/assets/291333/6bf78ff9-ce33-4b7f-893d-94f1b1eb6b2d)
  It's a bit of work in progress, has some hacks and requires cooperation with the modules that want to benefit from it though
- and yeah -- totally forgot about `HPI_LOGS` and broke it
  I guess it would be useful to keep it backwards compatible, I think it used to mean what `DEFAULT_LEVEL` means now? https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/logging.py#L49

 - For configuring loggers, could add some config settings I guess, but considering it's a bit up in the air, didn't want to change the configs structure etc. 
    So for now added a hack to load an `_init_hook.py` file which can be put next to the `config` module
    https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/core/__init__.py#L42-L47

    E.g. mine looks like this atm

    ```
    def configure_warnings() -> None:
        IGNORED = [
            ('script is DEPRECATED'   , 'instapexport.dal'),
            ('export_dir.*export_path', 'my.github.ghexport'),
            ('export_dir.*export_path', 'my.reddit.rexport'),
            ("Specifying modules' dependencies in the config.*is deprecated", 'my.instapaper'),
            ('modify your reddit config to look like', 'my.reddit'),
        ]
    
        import warnings
        for rgx, module_name in IGNORED:
            warnings.filterwarnings('ignore', '.*' + rgx + '.*', module=module_name)
    
    
    configure_warnings()
    
    
    def configure_env() -> None:
        import os
        os.environ['ENLIGHTEN_ENABLE'] = 'true'
    
    
    configure_env()
    ```


**cachew changes**
- I cleaned up logging a bit, now INFO is only for informing that cachew loads from the cache and writes to the cache. I figured all the stuff about specific hashes/disabled cachew etc ends up a bit verbose so kept it as DEBUG
- I added a bit of magic, so it automatically uses the logger of the module it's caching, unless you've explicitly passed a logger to cachew https://github.com/karlicoss/cachew/blob/b867166691572ece8f2bc3b671f8e24e4a6ce162/src/cachew/__init__.py#L752-L762
 
As an example, rescuetime module doesn't pass explicit logger, but cachew just reuses `my.rescuetime` logger
https://github.com/karlicoss/HPI/blob/c283e542e3457ecd778fb09e54e725d67104a49a/my/rescuetime.py#L31-L32

```
$ hpi query -s my.rescuetime.entries >/dev/null
[INFO    2023-09-06 23:06:00,710 my.rescuetime __init__.py:1005] my.rescuetime:entries: loading 685558 objects from cachew (sqlite /home/karlicos/.cache/my/my.rescuetime:entries)
```

(the `__init__.py` is kinda confusing though -- it comes from cachew, not my.rescuetime)

If you run it like this

```
LOGGING_LEVEL_my_rescuetime=DEBUG hpi query -s my.rescuetime.entries >/dev/null
```
you'll get the full cachew logs as well

```
LOGGING_LEVEL_my_rescuetime=DEBUG hpi query -s my.rescuetime.entries >/dev/null
[DEBUG   2023-09-06 23:09:44,218 my.rescuetime __init__.py:793 ] [my.rescuetime:entries] using inferred type typing.Union[rescuexport.dal.Entry, Exception]
[DEBUG   2023-09-06 23:09:44,219 my.rescuetime __init__.py:958 ] using /home/karlicos/.cache/my/my.rescuetime:entries for db cache
[DEBUG   2023-09-06 23:09:44,224 my.rescuetime __init__.py:962 ] new hash: {"cachew": "0.12.1.dev9+dirty", "schema": "[Column('_cachew_union_repr__Entry_is_null'
...
```

The downside I guess is that you'll get all DEBUG logs from `my.rescuetime` if there are any, but I figured it's a reasonable compromise to simplify things. 


Happy to get your input and suggestions!

**Comment by seanbreckenridge** [comment]
The automatic using of logger with cachew is really cool, I've always been a bit annoyed I had to create a logger just to ass it to cachew :+1: 

On the `import my._init_hook`, yeah, nice to separate that from the config I suppose. I've always had the `import my.config` at the top of every module, and if I want to do some global config I just do it in there. So having the config import at the top just handles any custom config as a sort of side effect, but that may not be obvious to someone looking at the code for the first time.

I dont think I'll document the init hook (yet), will try it myself and see if there are any rough edges first



> ENLIGHTEN_ENABLE

Wow, will definitely enable for myself, looks great :)

> I think it used to mean what DEFAULT_LEVEL means now

Yeah, I think so. Thanks for the explanations on the logging, about what I made sense of from reading it over. Can make a PR to maintain backwards comparability and add some docs (in `MODULE_DESIGN.org` maybe?)

**Comment by karlicoss** [comment]
>  I've always been a bit annoyed I had to create a logger just to ass it to cachew

I think a caveat is that you'll still need to create a logger (even if you don't pass it to cachew explicitly), otherwise it falls back on the default `cachew`'s logger.  https://github.com/karlicoss/cachew/blob/b867166691572ece8f2bc3b671f8e24e4a6ce162/src/cachew/__init__.py#L759
Was thinking about creating a logger automatically, but feels a bit intrusive, so not sure about it yet

> I dont think I'll document the init hook (yet), will try it myself and see if there are any rough edges first

Yep agreed -- I mainly wanted it to suppress some warnings at first, but later found out that warnings might actually need some rethinking completely because they are a bit messed up in python
https://memex.zulipchat.com/#narrow/stream/284580-python/topic/duplicate.20warnings.20in.20python

> Can make a PR to maintain backwards comparability and add some docs (in MODULE_DESIGN.org maybe?)

Sounds good, will appreciate it!

**Comment by seanbreckenridge** [comment]
resolved by #307
